bitbots_convenience_frames/src/convenience_frames.cpp

```diff
@@ -0,0 +1,209 @@
+      double y = non_support_foot_in_support_foot_frame.transform.translation.y / 2;
+      center_between_foot.setOrigin({0.0, y, 0.0});
+      center_between_foot.setRotation({0, 0, 0, 1});
+      tf2::Transform support_foot_tf;
+      tf2::fromMsg(support_foot.transform, support_foot_tf);
+      center_between_foot = support_foot_tf * center_between_foot;
+      approach_frame.pose.position.y = center_between_foot.getOrigin().y();
+      // z at ground leven (support foot height)
+      approach_frame.pose.position.z = support_foot.transform.translation.z;
+
+      // roll and pitch of support foot
+      double roll, pitch, yaw;
+      tf2::Quaternion quat;
+      fromMsg(support_foot.transform.rotation, quat);
+      tf2::Matrix3x3(quat).getRPY(roll, pitch, yaw);
+      // yaw of front foot
+      yaw = tf2::getYaw(front_foot.transform.rotation);
+
+      // pitch and roll from support foot, yaw from base link
+      tf2::Quaternion rotation;
+      rotation.setRPY(roll, pitch, yaw);
+      approach_frame.pose.orientation = tf2::toMsg(rotation);
+
+      // set the broadcasted transform to the position and orientation of the base footprint
+      tf_.header.stamp = ros::Time::now();
+      tf_.header.frame_id = "base_link";
+      tf_.child_frame_id = "approach_frame";
+      tf_.transform.translation.x = approach_frame.pose.position.x;
+      tf_.transform.translation.y = approach_frame.pose.position.y;
+      tf_.transform.translation.z = approach_frame.pose.position.z;
+      tf_.transform.rotation = approach_frame.pose.orientation;
+      broadcaster_.sendTransform(tf_);
+    } catch (...) {
+      continue;
+    }
+    r.sleep();
+  }
+}
+
+void ConvenienceFramesBroadcaster::supportFootCallback(const std_msgs::Char::ConstPtr &msg) {
+  got_support_foot_ = true;
+  is_left_support = (msg->data == 'l');
+}
+
+void ConvenienceFramesBroadcaster::ballCallback(const humanoid_league_msgs::BallRelative::ConstPtr &msg) {
+  publishTransform(msg->header.frame_id, "ball",
+                   msg->ball_relative.x, msg->ball_relative.y, msg->ball_relative.z);
+}
+
+void ConvenienceFramesBroadcaster::goalCallback(const humanoid_league_msgs::GoalRelative::ConstPtr &msg) {
+  publishTransform(msg->header.frame_id, "left_post",
+                   msg->left_post.x, msg->left_post.y, msg->left_post.z);
+  publishTransform(msg->header.frame_id, "right_post",
+                   msg->right_post.x, msg->right_post.y, msg->right_post.z);
+}
+
+void ConvenienceFramesBroadcaster::goalPartsCallback(const humanoid_league_msgs::GoalPartsRelative::ConstPtr &msg) {
+  for (long i = 0; i < msg->posts.size(); i++) {
+    publishTransform(msg->header.frame_id,
+                     "post_" + std::to_string(i),
+                     msg->posts[i].foot_point.x,
+                     msg->posts[i].foot_point.y,
+                     msg->posts[i].foot_point.z);
+  }
+}
+
+void ConvenienceFramesBroadcaster::publishTransform(std::string header_frame_id,
```

Comment:
If I see it correctly the method `publishTransform` could be made `static` right?

Reply:
I don't think so. broadcaster_ is needed